internal/controllers/upgrade_handler.go

```diff
@@ -0,0 +1,164 @@
+	} else if cmp == -1 {
+		return false, fmt.Errorf("template version (%v) is lower then ManagedCluster version (%v)",
+			templateVersion, managedClusterVersion)
+	}
+	return false, nil
+}
+
+func (t *provisioningRequestReconcilerTask) handleUpgrade(
+	ctx context.Context, renderedClusterInstance *siteconfig.ClusterInstance) (ctrl.Result, error) {
+	t.logger.InfoContext(
+		ctx,
+		"Start handleUpgrade",
+	)
+	clusterTemplate, err := t.getCrClusterTemplateRef(ctx)
+	if err != nil {
+		return requeueWithError(fmt.Errorf("failed to get clusterTemplate: %w", err))
+	}
+	ibgu, err := utils.GetIBGUFromUpgradeDefaultsConfigmap(
+		ctx, t.client, clusterTemplate.Spec.Templates.UpgradeDefaults,
+		clusterTemplate.Namespace, utils.UpgradeDefaultsConfigmapKey,
+		renderedClusterInstance.Spec.ClusterName, renderedClusterInstance.Namespace)
+	if err != nil {
+		return requeueWithError(fmt.Errorf("failed to generate IBGU for cluster: %w", err))
+	}
+	err = t.client.Get(ctx, types.NamespacedName{Name: ibgu.Name, Namespace: ibgu.Namespace}, ibgu)
+	if err != nil && errors.IsNotFound(err) {
+		if err := utils.CreateK8sCR(ctx, t.client, ibgu, t.object, utils.UPDATE); err != nil {
+			return requeueWithError(fmt.Errorf("failed to create IBGU: %w", err))
+		}
+
+		t.logger.InfoContext(
+			ctx,
+			fmt.Sprintf(
+				"Upgrade initiated. Created IBGU %s in the namespace %s",
+				ibgu.GetName(),
+				ibgu.GetNamespace(),
+			),
+		)
+
+		utils.SetStatusCondition(&t.object.Status.Conditions,
+			utils.PRconditionTypes.UpgradeCompleted,
+			utils.CRconditionReasons.InProgress,
+			metav1.ConditionFalse,
+			"Upgrade is initiated",
+		)
+		if err := utils.UpdateK8sCRStatus(ctx, t.client, t.object); err != nil {
+			return requeueWithError(fmt.Errorf("failed to update ClusterRequest CR status: %w", err))
+		}
+		return requeueWithMediumInterval(), nil
+	} else if err != nil {
+		return requeueWithError(fmt.Errorf("error getting IBGU: %w", err))
+	}
+	if IsIBGUProgressing(ibgu) {
+		utils.SetStatusCondition(&t.object.Status.Conditions,
+			utils.PRconditionTypes.UpgradeCompleted,
+			utils.CRconditionReasons.InProgress,
+			metav1.ConditionFalse,
+			"Upgrade is in progress",
+		)
+		t.logger.InfoContext(
+			ctx,
+			"Wait for upgrade to be completed",
+		)
+		if err := utils.UpdateK8sCRStatus(ctx, t.client, t.object); err != nil {
+			return requeueWithError(fmt.Errorf("failed to update ClusterRequest CR status: %w", err))
+		}
+		return requeueWithMediumInterval(), nil
+	}
+
+	if failed, message := isIBGUFailed(ibgu); failed {
+		utils.SetStatusCondition(&t.object.Status.Conditions,
+			utils.PRconditionTypes.UpgradeCompleted,
+			utils.CRconditionReasons.Failed,
+			metav1.ConditionFalse,
+			message,
+		)
+	} else {
+		utils.SetStatusCondition(&t.object.Status.Conditions,
+			utils.PRconditionTypes.UpgradeCompleted,
+			utils.CRconditionReasons.Completed,
+			metav1.ConditionTrue,
+			"Upgrade is completed",
+		)
+		err := t.client.Delete(ctx, ibgu)
```

Comment:
I think deleting the IBGU should move at the end. If the delete fails, then the ProvisioningRequest status is not updated to UpgradeCompleted True, even if it should.

Reply:
I've re-arranged this function quite a bit. I will see if I can address this in my version without making it too complicated.